logutil/logutil.go

```diff
@@ -0,0 +1,132 @@
+// Package logutil provides various logging utilities that are meant to expand the capabilities of [log/slog].
+//
+// [PrettyHandler] is a [slog.Handler] that outputs logs in a text format similar to [slog.TextHandler] but
+// with pretty formatting and colours. It is intended for use in CLIs to make easy to read logs for users.
+//
+// [MultiHandler] is a [slog.Handler] that allows a single log record to be processed by multiple handlers.
+// It is akin to [io.MultiWriter]. Each handler has the ability to customize its behaviour.
+package logutil
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"runtime"
+	"sync"
+
+	"github.com/TouchBistro/goutils/progress"
+)
+
+// WriterVar is a io.Writer variable, to allow a Handler writer to change dynamically.
+// It implements io.Writer as well as a Set method, and is safe for use by multiple goroutines.
+//
+// A WriterVar must not be copied after first use.
+//
+// The zero value LevelVar is a no-op writer that discards all data written to it (similar to io.Discard).
+type WriterVar struct {
+	w  io.Writer
+	mu sync.Mutex
+}
+
+// NewWriterVar creates a new WriterVar with the given writer.
+func NewWriterVar(w io.Writer) *WriterVar {
+	return &WriterVar{w: w}
+}
+
+// Set sets the underlying writer.
+func (v *WriterVar) Set(w io.Writer) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	v.w = w
+}
+
+func (v *WriterVar) Write(p []byte) (int, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	// Handle zero value case, if no writer just no-op.
+	if v.w == nil {
+		return len(p), nil
+	}
+	return v.w.Write(p)
+}
+
+// LogWriter returns an io.Writer that can be used to write arbitrary text to the logger.
+// logger is used to log an error if one occurs.
+//
+// It is the caller's responsibility to close the returned io.WriteCloser in order
+// to free resources.
+func LogWriter(logger progress.Logger, level slog.Level) io.WriteCloser {
+	pr, pw := io.Pipe()
+	var logFunc func(string, ...any)
+	switch level {
+	case slog.LevelDebug:
+		logFunc = logger.Debug
+	case slog.LevelInfo:
+		logFunc = logger.Info
+	case slog.LevelWarn:
+		logFunc = logger.Warn
+	case slog.LevelError:
+		logFunc = logger.Error
+	default:
+		// See if the logger has a Log method that can be passed a level.
+		type withlog interface {
+			Log(context.Context, slog.Level, string, ...any)
+		}
+		wl, ok := logger.(withlog)
+		if !ok {
+			panic(fmt.Errorf("logutil.LogWriter: unsupported level %s(%d)", level.String(), level))
```

Comment:
I wonder instead of this hack would it make more sense to explicitly support arbitrary log levels in `progress.Logger`? That would mean coupling it to `slog`. I was kinda hoping to avoid that but I wonder if we just have the illusion of it being decoupled and really it would be better to just say this has to work with slog?